Add support for OAuth client credentials

clstokes opened the pull request:
**What this PR does / why we need it**:

Allows configuring the tailscale client with OAuth Client ID and OAuth Client. On instantiation the Client will generate a new API key against the Tailscale OAuth token endpoint.

**Which issue this PR fixes**: 

Fixes #42

knyar commented on tailscale/client.go:
```diff
@@ -45,6 +47,25 @@ type (
+// OAuth Credentials. The Client will generate a new API key against the Tailscale OAuth token endpoint.
+// If the API key expires, it will not be automatically renewed by the Client.
+// Additional options can be provided, see ClientOption for more details.
+func NewClientFromOAuth(oauthClientID, oauthClientSecret, tailnet string, options ...ClientOption) (*Client, error) {
```
Could you please make this accept a `ctx.Context` instead of using `context.Background`?

knyar commented on tailscale/client.go:
```diff
@@ -45,6 +47,25 @@ type (
 
+// NewClientFromOAuth returns a new instance of the Client type that will perform operations against a chosen tailnet using
+// OAuth Credentials. The Client will generate a new API key against the Tailscale OAuth token endpoint.
+// If the API key expires, it will not be automatically renewed by the Client.
```
Is there any reason why the token is not renewed? From a user's point of view I would expect this to be handled by the library. Reading the docs, it seems easy to achieve via an `http.Client` [provided](https://pkg.go.dev/golang.org/x/oauth2/clientcredentials#Config.Client) by `oauth2/clientcredentials`.

clstokes replied:
Will look into this.

clstokes replied:
Per an internal discussion, the Tailscale OAuth token endpoint does not return a `refresh_token` and clients must use OAuth client credentials to get a new API key. This means the auto-refresh functionality from `oauth2/clientcredentials` cannot be used. 

@knyar given this, do you still think users of `tailscale-client-go` will expect the client to to auto-renew the api key with the originally supplied OAuth credentials?

willnorris replied:
The way the golang.org/x/oauth2 library is intended to work is that it gives you an http.Client that you should use to make outbound calls.  That client with transparently grab new tokens as needed.  But what's happening below is that you're grabbing a single access token, and passing that to NewClient().  Instead, NewClient needs to take an http.Client (or expose a way to set the underly http.Client).  

For example, see [google/go-github](https://github.com/google/go-github#authentication) as an example.  The example there uses a static token (which would be equivalent for traditional API keys for us, or it can taken any other oauth2 client that will handle token refresh.

clstokes replied:
My local testing with the Client from golang.org/x/oauth2 showed that it was unable to transparently grab new tokens without a refresh token. I never pushed this code based on my tests failing, but you can see my attempt at https://gist.github.com/clstokes/c56477accb1076356ae60c4529671110#file-client-go-L93.

clstokes replied:
My latest push adds token renewal.

willnorris replied:
you shouldn't need to do token renewal yourself.  The gist above should just be:

```go
    http:    oauthConfig.Client(ctx),
```

You don't need a new Client constructor either, just add a ClientOption that sets the underlying http.Client.  Here's what I mean (along with a simple example script the demonstrates it working):

https://github.com/tailscale/tailscale-client-go/commit/088bc8a613490528459730cc3cec0f5ebc94b79b

willnorris replied:
eventually, it'd be great to get rid of the `apiKey` parameter to `NewClient` and replace it with:

``` go
func WithAPIKey(key string) ClientOption {
    ...
}
```

But that's a breaking change so perhaps something that will have to come later.

knyar replied:
I've been thinking about this too. I don't think we have a policy on breaking changes, and in practice this library is primarily used by our [Terraform provider](https://github.com/tailscale/terraform-provider-tailscale), so making a breaking change in lockstep with the provider might be acceptable.

If we decide to keep `NewClient` as is, we should probably have the new function be more generic rather than oauth2-specific. For example, we could have:
- `func NewClientContext(ctx context.Context, tailnet string, options …ClientOption) (*Client, error)`
- `func WithBaseURL(baseURL string) ClientOption`
- `func WithAPIKey(apikey string) ClientOption`
- `func WithOauthClient(oauthClientID, oauthClientSecret string, oauthScopes []string) ClientOption`

As part of this `NewClient` can just call `NewClientContext` with `WithAPIKey` and have no code duplication.

Any thoughts?

knyar commented on tailscale/client.go:
```diff
@@ -45,6 +47,25 @@ type (
+// If the API key expires, it will not be automatically renewed by the Client.
+// Additional options can be provided, see ClientOption for more details.
+func NewClientFromOAuth(oauthClientID, oauthClientSecret, tailnet string, options ...ClientOption) (*Client, error) {
+	var oauthConfig = &clientcredentials.Config{
```
Shouldn't `scopes` be specified as well?

willnorris replied:
Tailscale OAuth clients already have a set of scopes configured on them, and by default access tokens will have those same scopes.  You only need to specify scopes here if you want to further limit the scopes applied, but I'm not sure if that will be very common.

clstokes replied:
My latest push adds support for passing in scopes.

willnorris reviewed (changes requested):
explicitly marking 'request changes' based on earlier comments about using the oauth2-provided `http.Client`

clstokes commented:
@knyar @willnorris I believe this is ready for another review.

knyar commented on tailscale/client.go:
```diff
@@ -45,6 +53,45 @@ type (
+	}
+
+	// return an error message at client initialization if oauth credentials fail instead of waiting for first request
+	_, err = c.apiKeyFromOAuth(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("error getting API key from OAuth token endpoint: %w", err)
+	}
```
I think we'd want to do this after applying `options`, just before returning from this function.

clstokes commented:
@knyar @willnorris, mind giving this another look?

willnorris reviewed (approved):
Looks great!  Thanks for carrying this all the way through.

knyar reviewed (approved):
Thank you!

clstokes commented:
For those following along I rebased and squashed my commits here to sign and signoff on the commits to pass the commit signature verification requirement.
